### DISCUSSION TITLE
Best way to handle large map of values.

### DISCUSSION THREAD

**Discussion by daprahamian** [pmndrs/jotai#1942] [Q&A]
I'm got a collection of values that need to be referenced based on ID, but also need to be ordered. A simplified version:

```ts
interface Obj {
  id: string;
  // Insert a bunch of properties here
}
interface Collection {
  order: string[];
  values: Record<string, Obj>;
}
```

I'm currently dealing with a few constraints that are puzzling me, that being:

- How do I easily make it so a component can directly read and write to a sub-object while only subscribing to changes from that sub-object
- How do I make it easy to, given an id, verify whether or not there is already an object with that id?
- How do I make all the typings work out nicely?
- How do I make it easy to reorder the objects?
- How do I make it easy to serialize / deserialize to/from JSON?

So far I've tried:

## 1: Atoms in Atoms

```ts
const listAtom = atom<string[]>([]);
const mapAtom = atom<Record<string, Obj>>({});
```

this is good for updating to the list, however, I have not found a good way of getting a specific atom from within `mapAtom`. The basic React hook looks like:

```ts
function useObj(id) {
  const map = useAtomValue(mapAtom);
  return useAtom(map[id]);
}
```

which pulls in the whole map.

## 2. Atom Families

```ts
const listAtom = atom<string[]>([]);
const objAtomFamily = atomFamily(id => atom({id}));
```

This works great for being able to get the atom from anywhere and for reordering, but:

1. There is no way to verify whether or not a given id is already in use
2. The "MEMORY LEAKS!" disclaimer on atomFamily is kind of scary.

## 3. Optics

```ts
const collectionAtom = atom<Collection>({order: [], values: {} });
const listAtom = focusAtom(collectionAtom, o => o.prop('order'));
const mapAtom = focusAtom(collectionATom, o => o.prop('values'));
```

I feel like the solution might be somewhere in optics, but I can't figure out how to easily access a subvalue of `mapAtom`.

Any advice?

**Comment by dai-shi** [comment]
> How do I make it easy to serialize / deserialize to/from JSON?

This requirement would suggest one big atom. So, "3. Optics" would be good. But, you may not need optics. Just deriving atoms might work hopefully.

> I can't figure out how to easily access a subvalue of mapAtom.

Can show a very small example?

**Reply by daprahamian** [reply]
I guess looking into it, the answer is probably to do something like:

```ts
const collectionAtom = atom<Collection>({order: [], values: {} });
const listAtom = focusAtom(collectionAtom, o => o.prop('order'));
const mapAtom = focusAtom(collectionAtom, o => o.prop('values'));

const objAtomFamily = atomFamily(id => focusAtom(mapAtom, o => o.prop('id'))
```

would that be accurate?

**Reply by daprahamian** [reply]
I think the main thing I am struggling with is how to delete a prop from `mapAtom`. I can set it to `undefined`, but I want the actual property deleted.

**Reply by dai-shi** [reply]
> would that be accurate?

I think you mean this?
```js
const objAtomFamily = atomFamily((id) => atom((get) => get(mapAtom)[id]))
```
It creates read-only atoms, but you can make it to create read-atoms too. Not sure if focusAtom can do with types.

> delete a prop from mapAtom. 

You can do that. But, I'm curious how you are doing it currently. Can share a code snippet?

**Reply by blissdev** [reply]
How do you make it create write atoms? I'm getting a bit stuck with this piece of `atomFamily`.

Is this it, or is there more to it?

```js
export const residenceUsageAtom = atom({});
export const residenceUsageFamily = atomFamily((chunkId) =>
    atom(
        (get) => get(residenceUsageAtom)[chunkId],
        (get, set, arg) => {
            const prev = get(residenceUsageAtom);
            return {
                ...prev,
                [chunkId]: arg,
            };
        }
    )
);
```

**Reply by daprahamian** [reply]
So making it create with write atoms I found two approaches:

```ts
const mapAtom = atom({});
const objAtom = atomFamily(id => focusAtom(mapAtom, o => o.prop(id)));

// Approach 1: Modify the root
const addObjAtom = atom(null, (get, set, obj) => {
  set(mapAtom, O.set(O.optic().prop(obj.id))(obj));
});

// Approach 2: Instantiate atomFamily and immediately set:
const addObjAtom2 = atom(null, (get, set, obj) => {
  set(objAtomFamily(obj.id), obj);
});
```

The area I was struggling with was deleting atoms. I was hoping for something elegant with optics like:

```ts
const mapAtom = atom({});
const objAtom = atomFamily(id => focusAtom(mapAtom, o => o.prop(id)));
const deleteObjAtom = atom(null, (get, set, id) => {
  set(mapAtom, O.remove(O.optic().prop(id))
});
```

but `remove` only works on arrays, and instead leaves behind a `Symbol(__remove_me__)` as the value when you try to use it on an object. I wound up having to do:

```ts
function removeMapEntry(key) {
  return function remove(o) {
    const {[key]: _, ...acc} = o;
    return acc;
  };
}

const mapAtom = atom({});
const objAtom = atomFamily(id => focusAtom(mapAtom, o => o.prop(id)));
const deleteObjAtom = atom(null, (get, set, id) => {
  set(mapAtom, removeMapEntry(id));
});
```

So I guess what I wish was that there was an equivalent to `O.remove` that unset a property on an object. And even neater, if there was an elegant way to make the removal op work on the property lens instead of on the root object. Something similar to `atomWithReset`:

```ts
import {focusAtom, UNSET} from 'jotai-optics'

const mapAtom = atom({});
const objAtom = atomFamily(id => focusAtom(mapAtom, o => o.prop(id)));

const deleteObjAtom = atom(null, (get, set, id) => {
  set(objAtomFamily(obj.id), UNSET);
});
```

**Reply by dai-shi** [reply]
Not 100% following, so there might be a misunderstanding of mine, but I'd probably avoid using focusAtom if that doesn't work well. We can technically do what focusAtom can do manually.

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/discussions/1942#discussioncomment-5958953

@blissdev Yes, that's the right way:

```js
const writableAtomFamily = atomFamily((param) => atom(
  (get) => ...,
  (get, set, arg) => ...
))
```

**Comment by lveillard** [comment]
Hello! I'm a bit confused with the writtable familyAtoms. Some concepts are so intuitive and clean, and others are killing me haha.

isn't there a way to just add a default value to all the generated atoms of a family, and a setter as well?

```
export const residenceUsageAtom = atom({});
export const residenceUsageFamily = atomFamily((chunkId) =>
    atom(
        (get) => get(residenceUsageAtom)[chunkId],
        (get, set, arg) => {
            const prev = get(residenceUsageAtom);
            return {
                ...prev,
                [chunkId]: arg,
            };
        }
    )
);
```
👆 This is forcing us to use a secondary atom and store all the values in that store. Ideally I would love to be able to do something like this 

👇 What i'm trying to achieve
```js

export const selectedIdsInPagesFamily= atomFamily((page) =>
    atom(
        (get,current) =>  current ? current : 'default'
        (get, set, current) => {
             // do some changes to other atoms
             set( ... )
             // 
             return arg 
            };
        }
    )
);
```

With this we could use it in a component somethibg like this:

```js
[selectedIdsInPage, setSelectedIdsInPage] = useAtom(selectedIdsInPagesFamily('myPage`);
const handleArrayChangesInTheSelext = (values) => {setSelectedIdsInPage(values)}
```

But I'm struggling to:
1) Build a default value for the getter
2) Send the current value in the getter
3) Make the atom writtable and just update itself with the value it receives, overriding the default value

**Reply by dai-shi** [reply]
It sounds like you have a misconception about the capability of atomFamily. It's just a Map and doesn't do other jobs.

https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035 has a simplified code of atomFamily
```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```

**Reply by lveillard** [reply]
I see! I'm loving jotai so far but struggling a bit with some concepts.

Is there another way to have an atomFactory to add dynamically atoms of the same format? I have a similar use case to daprahamian

For instance, a really basic use case:

imagine fetching from db a list of employees, each one with the fields "birthday" (editable) and "age" (computed, depends on birthday). 

I'm looking for a way to:
- batch-update/add a bunch of employees with their birthdays
- Be able to fetch them by id or id[] later

I think i'm overusing this structure 😅
```
// updating the values cache when the units cache is updated
 newEmployeesBatch.forEach((employee) => {
            set(valuesBirthdayFamily(employee.birthday), employee.id);
          });
````

**Reply by dai-shi** [reply]
Having everything in one big atom is not that bad and you can derive atoms from it to optimize rendering.
I would explore the possibility to use the atoms-in-atom pattern: https://jotai.org/docs/guides/atoms-in-atom

**Reply by lveillard** [reply]
Got it! Atoms-in-atom don't look like being adapted to my use case as i need to send a parameter somehow to get / update the right value.

In the family side, i'm exploring these two patterns, not sure which one will be more performant or if there really is a way to do it with atom-in-atoms equivalent.

**Pattern A: Storing the data in the atoms**
```js
export const valueIdsAtom = atom(new Set<string>());

export const valueAtomFamily = atomFamily(
  ([employeeId, propertyId]) =>
    atom(undefined, (get, set, update) => {
      const compositeId = `${employeeId}::${propertyId}`;
      // get the current value
      const newValue =
        typeof update === 'function'
          ? update(get(valueAtomFamily([employeeId, propertyId])))
          : update;

      /// update the localValueIdsAtom
      const currentIds = get(valueIdsAtom);
      const newIds = new Set([...currentIds, compositeId]);
      set(valueIdsAtom, newIds);

      // Set the new value to the atom.
      set(valueAtomFamily([employeeId, propertyId]), newValue);
    }),
  deepEqual
);
```

**Pattern B: Storing the data in a Map**
```js
export const valuesAtom = atom(new Map());

export const valueAtomFamily = atomFamily(
  ([employeeId, propertyId]) =>
    atom(undefined, (get, set, update) => {
      const compositeId = `${employeeId}::${propertyId}`;
      // get the current value
      const newValue =
        typeof update === 'function'
          ? update(get(valueAtomFamily([employeeId, propertyId])))
          : update;

      const prev = get(valuesAtom);
      const updatedValues = doingSomeStuffToTheCurrentCompositeId(newValue)
      const next = new Map(updatedValues);
      set(valuesAtom, next);

    }),
  deepEqual
);
```

**Reply by dai-shi** [reply]
The difference is Set or Map?

AtomFamily wouldn't be necessary if you are just creating a write-only atom. (You could use it though.)
You can just define one write-only atom and receive ids as arguments.

**Reply by lveillard** [reply]
Yes, either Set for the keys iteration+ familyAtom storing the data, or Map for keys&data and family atom to edit / get data.

For the atom-in atom alternative, the problem is that the param is not a state, in the same component i might dynamically fetch some data and require to use some of that data to focus on a particular key of the map. Or two different values for the same  simultaneously in two different pars of the code.

```js
export const relationsMap = atom(new Map());

export const setRelationAtom = atom(null, (get, set, relation: Relation) => {
  const currentMap = get(relationsMap);
  const newMap = new Map(currentMap);
  newMap.set(relation.id, relation);
  set(relationsMap, newMap);
});

export const getRelationByIdAtom = atom((get) => (id: string) => {
  const currentMap = get(relationsMap);
  return currentMap.get(id);
});
```

This for instance(Map version) looks good but then when I call:

` const id1 = useAtomValue(getRelationByIdAtom)('1');`
 
The component does not re-render so i guess atom-in-atom does not work for this use case or i'm doing something wrong 😓

**Reply by dai-shi** [reply]
What we often do is to create a new atom in a component.
```js
const value = useAtomValue(useMemo(() => atom((get) => get(bigObjAtom)[id]), [id]))
```

**Reply by lveillard** [reply]
Ok this works, even with Map instead of objects and it only gets re-rendered when the selected value does. Black magic.

Are the docs open source? Can I open a PR to add an example in the atomFamily docs so people know about this alternative?

**Reply by dai-shi** [reply]
Yes, feel free to open PRs to docs and we will review them.

**Comment by songkeys** [comment]
> How do I make it easy to serialize / deserialize to/from JSON?

This is what I want. In my case, the JSON comes from network request on initialization and only small parts of it will get updated in the following polling network requests.

The network request is made in a hook in a top component `<Root>`:

```
<Root>
  {veryComplexTree.map(...)}
</Root>
```

Storing the huge JSON in one atom is obviously expensive for rendering. So I tried to split the atoms.

But if I choose to create new atoms in components, when a new network command comes in (like `{ "action": "update tree['a']['b']['c']" }`), I might need to add event listeners between each tree item and the network requester to notify the target item to update its own atom? This is too complex to me.

What is more intuitive to me is updating each atom in the top root network requester, i.e., manage atoms in one place. So this brings the first question: how to do json -> atoms in atoms in the initial rendering more easily?

I tried `splitAtom` but it looks like it only works for top level array but not deep nested array. Splitting from top to bottom to make everything atoms seems impossible.

What is the best way to solve this?

----

edit: I wish there was an API called `atomize` which serialize JSON into nested atoms automatically and dynamically. Is this possible?

```ts
const nestedAtom = atomize({
  "users": [
    { "username": "jason", "age": 17 },
    { "username": "tom", "age": 21 },
  ]
})

// ==

const nestedAtom = {
  users: atom([
    atom({ username: atom("jason"), age: atom(17) }),
    atom({ username: atom("tom"), age: atom(21) }),
  ])
}
```

**Reply by dai-shi** [reply]
> I wish there was an API called atomize which serialize JSON into nested atoms automatically and dynamically. Is this possible?

Why not?
```js
const atomize = (x) =>
  Array.isArray(x) ? x.map(atomize) :
  typeof x === 'object' ? Object.fromEntries(Object.entries(x).map(([k, v]) => [k, atomize(v)])) :
  atom(x);
```

**Reply by lveillard** [reply]
Interesting concept, how would that work to read/edit them? 🤔

**Reply by dai-shi** [reply]
https://jotai.org/docs/guides/atoms-in-atom might help?
